src/app/shared/directives/stretch-table/stretch-table.directive.ts

```diff
@@ -0,0 +1,99 @@
+import { AfterViewChecked, AfterViewInit, ComponentFactoryResolver, Directive, ElementRef, HostListener, Renderer2, ViewContainerRef } from '@angular/core';
+import { StretchCellComponent } from '../../components/stretch-cell/stretch-cell/stretch-cell.component';
+
+@Directive({
+  selector: '[appStretchTable]'
+})
+export class StretchTableDirective implements AfterViewInit{
+  private selectedTh!: any;
+  private tableContainerWidth!: any;
+  private mouseMoveFunc!: any;
+
+  constructor(
+    private el: ElementRef,
+    private componentFactoryResolver: ComponentFactoryResolver,
+    private renderer: Renderer2,
+    private viewContainerRef: ViewContainerRef
+  ) { }
+  ngAfterViewInit(): void {
+    setTimeout(() => this.setResizeFunction(), 100);
+  }
+
+  private setResizeFunction(){
+    let THs = this.el.nativeElement.getElementsByTagName('th');
+    
+    for(let i = 0; i < THs.length - 1; i++){
+      if(THs[i + 1].classList.contains('mat-table-sticky') 
+        && THs[i + 1].classList.contains('mat-table-sticky-border-elem-right')) {
+        break; 
+      }
+
+      const componentFactory = this.componentFactoryResolver.resolveComponentFactory(StretchCellComponent);
+      let componentRef = this.viewContainerRef.createComponent(componentFactory);
+      let content = THs[i].childNodes[0];
+
+      this.renderer.removeChild(THs[i], content);
+      componentRef.instance.insertNode(content);
+      
+      this.renderer.appendChild(THs[i], componentRef.location.nativeElement);
+    }
+
+    for (let i = 0; i < THs.length; i++) {
+      THs[i].style.width = `${THs[i].offsetWidth}px`;
+    }
+  }
+  
+  
+  @HostListener('mousedown',['$event']) onMouseDown(event: any){
+    if(!event.target.closest(".resize-border")){
+      return;
+    }
+
+    this.selectedTh = event.target.closest("th");
+    this.tableContainerWidth = this.selectedTh.closest('.table-container').offsetWidth;
+    this.mouseMoveFunc = this.changeWidth.bind(this);
+
+    document.addEventListener('mouseup', this.onUpMouse.bind(this), {once: true})
+    document.addEventListener('mousemove', this.mouseMoveFunc);
+
+    document.querySelector('body')!.style.userSelect = 'none';
+    document.querySelector('body')!.style.pointerEvents = 'none';
```

Comment:
save this value document.querySelector('body') to constant and then use, so we will not try to find the same element twice. The same thing for other cases in the code